### DISCUSSION TITLE
Images that upload successfully via `createBlob` will then fail to load, resulting in a 422 instead

### DISCUSSION THREAD

**Discussion by davidcelis** [bluesky-social/atproto#1777] [Dev Debugging] [answered]
**Describe the bug**

I primarily post small notes to my own website, then syndicate them to both Mastodon and Bluesky. Like Mastodon/Bluesky/Twitter, my small notes can have up to four attached images; when I syndicate these posts, I'll process the images based on the platform's limits. For Mastodon, that means I need to downsize them to have fewer than 33,177,600 pixels and compress them to be smaller than 16 MB. For Bluesky, I compress them until they're smaller than 976.56 KB.

I don't know for _sure_ if it's related but, ever since I upgraded from an iPhone 14 Pro Max to an iPhone 15 Pro Max, most of my images that I upload are unable to load in Bluesky's client applications and instead return a 422 from the CDN. I'm unable to debug this reasonably on my own, because the actual API calls to `/xrpc/com.atproto.repo.uploadBlob` succeed and return a blob object that I'm then able to embed in my post. It's only when the Bluesky client applications attempt to show these photos via its CDN that things go wrong.

At first, I thought it had to do with the image's orientation (images taken in landscape mode tended to fail to load, but images taken in portrait mode would). However, I've now seen at least one portrait image I uploaded now fail to load with the same issue. Here are a list of posts that display the problem:

* https://bsky.app/profile/davidcel.is/post/3kbjinep2f22w
* https://bsky.app/profile/davidcel.is/post/3kbgxklgjbs2w
* https://bsky.app/profile/davidcelis.bsky.social/post/3kbecvzojat2d
* https://bsky.app/profile/davidcel.is/post/3kbbm3coynf2p
* https://bsky.app/profile/davidcel.is/post/3kb723jewh42o
* https://bsky.app/profile/davidcel.is/post/3kb6wm2zdji2c
* https://bsky.app/profile/davidcelis.bsky.social/post/3kbbnuci3l32z
* https://bsky.app/profile/davidcelis.bsky.social/post/3kb6cuitrhq25

Meanwhile, the images in these recent posts all load, despite undergoing the exact same processing:

* https://bsky.app/profile/davidcel.is/post/3kbicj7fajs2h
* https://bsky.app/profile/davidcel.is/post/3kbfyjhf7wg2d
* https://bsky.app/profile/davidcel.is/post/3kbdhlxqdj42y
* https://bsky.app/profile/davidcel.is/post/3kb6hwfh6yg2v
* https://bsky.app/profile/davidcelis.bsky.social/post/3kbc4dth4kh2x
* https://bsky.app/profile/davidcelis.bsky.social/post/3kbc42b654j26

**To Reproduce**

I'm… not really sure of an easy way to reproduce this outside of using my actual code, which is a Rails app. My process is:

1. I upload images to my website directly from my phone (or, if I'm on my mac, from Photos.app)
2. If I uploaded an HEIC instead of a JPEG, I convert it to a JPEG using Vips
3. Then, I use Vips JPEG saver with lower and lower `Q`/`quality` values until the resulting image is tiny enough for Bluesky's <1MB limit.

**Expected behavior**

Images that successfully upload via `/xrpc/com.atproto.repo.uploadBlob` are _also_ able to display. If images are going to end up not being able to display, the `uploadBlob` method should fail.

**Additional context**

I think the same issue may have been reported previously in #1389, but it was closed by the author on the same day with no comment. If it's better to close this and re-open that report, then that makes sense!

**Comment by devinivy** [comment, marked as answer]
I think the issue here is that the image resizer has a default max resolution that it will process, which is 16.8 megapixels.  If you shrink the image down a bit, I think you'll be back in business!  Unfortunately I don't think we're in a great position to enforce this during `uploadBlob`, since it's not technically constrained to image uploads for the bluesky app.  I'll take a look into what we might be able to do, though, to give more proactive feedback about whether an image will be able to display in the app.

**Comment by davidcelis** [comment]
Ah, thanks a ton! I'll add that resolution limit on my side as soon as I can and start testing things out again 🙌

**Comment by davidcelis** [comment]
That did it! Thanks again for your help, @devinivy. I'll let you close this or keep it open if it would be a good place to track the issue of better feedback during or after uploads.

**Reply by bnewbold** [reply]
Moved to a "discussion" to keep the response around. Thanks for the well written initial report!
